lib/babe/epoch_handler.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+// The gossamer library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package babe
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
+)
+
+type handleSlotFunc = func(epoch, slotNum uint64, authorityIndex uint32, proof *VrfOutputAndProof) error
+
+var (
+	errEpochPast = errors.New("cannot run epoch that has already passed")
+)
+
+type epochHandler struct {
+	epochNumber uint64
+	firstSlot   uint64
+
+	constants *constants
+	epochData *epochData
+
+	// for slots where we are a producer, store the vrf output (bytes 0-32) + proof (bytes 32-96)
+	slotToProof map[uint64]*VrfOutputAndProof
+
+	handleSlot handleSlotFunc
+}
+
+func newEpochHandler(epochNumber, firstSlot uint64, epochData *epochData, constants *constants,
+	handleSlot handleSlotFunc, keypair *sr25519.Keypair) (*epochHandler, error) {
+	// determine which slots we'll be authoring in by pre-calculating VRF output
+	slotToProof := make(map[uint64]*VrfOutputAndProof)
+	for i := firstSlot; i < firstSlot+constants.epochLength; i++ {
+		proof, err := claimPrimarySlot(
+			epochData.randomness,
+			i,
+			epochNumber,
+			epochData.threshold,
+			keypair,
+		)
+		if err != nil {
+			if errors.Is(err, errOverPrimarySlotThreshold) {
+				continue
+			}
+			return nil, fmt.Errorf("error running slot lottery at slot %d: error %w", i, err)
+		}
```

Comment:
```suggestion
		if errors.Is(err, errOverPrimarySlotThreshold) {
				continue
	         } else if err != nil {
			return nil, fmt.Errorf("error running slot lottery at slot %d: error %w", i, err)
		}
```

Reply:
nit I would also suggest to remove the `error ` prefix before `%w`

Reply:
updated